### PR TITLE
Add Generate teams method to team model

### LINKED ISSUE #73: Add generate teams method to the team model
Add a "Generate Team" method to the team model to enable easy division of users into two teams.

### PR DESCRIPTION
**Pull request by rotem2022**
### What this PR does?

In this PR I added a method that gives an option to generate two teams based on an event.

### Notes for reviewer

I divided this PR into two commits

1. a stab commit witch contains files the changing depending on from #56 that should be deleted after
2. a commit for changes relevant for this PR

### Related Issue

Resolves #73

#### Test coverage

- [x] This PR includes new unit and integration tests to go with the code
  changes
- [ ] The changes in this PR do not require tests

#### Documentation

- [ ] Docs (e.g. `README`s) were updated in this PR
- [x] This PR does not require updating any documentation

### CODE REVIEW

**Comment by ygalblum** [comment]
@rotem2022 there's a conflict you need to fix

**Comment by rotem2022** [comment]
> @rotem2022 there's a conflict you need to fix

All done.

File: teams/models.py [hunk 1]
```diff
@@ -11,3 +11,17 @@ def save(self, *args, **kwargs):
...
+        users_events = UserEvent.objects.filter(eventID=event_id)
+        team_size = len(users_events) // 2
+        team1 = Teams(name="Team1")
+        team1.save()
+        team2 = Teams(name="Team2")
+        team2.save()
+        UserEvent.objects.filter(userID__in=[user_event.userID for user_event in users_events[:team_size]],
+                                 eventID=event_id).update(teamID=team1)
+        UserEvent.objects.filter(userID__in=[user_event.userID for user_event in users_events[team_size:]],
+                                 eventID=event_id).update(teamID=team2)
```
Comment (talko611) [inline teams/models.py:27]: You are assuming that the number of users in the event is divisible by 2. but it is not the case, there are users that won't be entered into either one of the teams.

**Review by talko611 — CHANGES_REQUESTED** [review]

File: teams/models.py [hunk 2]
```diff
@@ -11,3 +11,17 @@ def save(self, *args, **kwargs):
         if self.name in teams_names:
             raise IntegrityError
         super().save(*args, **kwargs)
+
+    @staticmethod
+    def generate_teams(event_id):
+        from event.models import UserEvent
```
[the remark this answers is not in the document]
Comment (rotem2022) [reply teams/models.py:17]: The import cannot be placed at the top because it would cause a circular dependency between the team and event imports.

File: teams/models.py (hunk 1 above)
Comment (rotem2022) [reply teams/models.py:27]: Actually, I am not making this assumptions. If the number of users in the event is not divisible by 2, then the second team will be larger. This can be seen in the test that has been conducted

**Review by SaarOrus — APPROVED** [review]

File: teams/models.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (talko611) [reply teams/models.py:17]: Ok

File: teams/models.py (hunk 1 above)
Comment (talko611) [reply teams/models.py:27]: Ok, sorry I missed that.
But then I don't understand the purpose of the team division if you sometimes get uneven teams.
I think you should divide the teams evenly and create another team with a special name and put there the participants that have no team

Comment (rotem2022) [reply teams/models.py:27]: I appreciate your suggestion, but in this case, having a "special" team for participants who don't fit evenly into two teams might create an unnecessary distinction between participants. By having the second team be larger, we can avoid this and still have a fair and functional team division. Additionally, having two teams of unequal size can create a competitive advantage for the smaller team, which can add an interesting element to the event. Overall, I believe that leaving the teams uneven in certain cases can still achieve the desired outcome without creating unnecessary complexity.

Comment (talko611) [reply teams/models.py:27]: Why is it an "unnecessary distinction"? this is equal to what happens in real life, you divide to equal teams and the leftovers are put in a special team. you can call this team "waiting" or "observers" to distinguish them from the regular teams.
I don't think somebody wishes to play a team sport with uneven teams and it misses the whole point of dividing into teams.

**Review by talko611 — DISMISSED** [review]

File: teams/tests.py [hunk 3]
```diff
@@ -1,13 +1,124 @@
 from django.core.exceptions import ValidationError
 from django.db.utils import IntegrityError
 from .models import Teams
+from event.models import UserEvent, Event
```
Comment (ygalblum) [inline teams/tests.py:4]: Please reorder the imports. Generally speaking there should be three sections: Python Standard, External and Internal. In this case, you don't import any standard libraries, the `pytest` and `django` imports are external and the rest is internal. In addition, make sure that each section is sorted alphabetically

File: teams/tests.py [hunk 4]
```diff
@@ -27,3 +138,21 @@ def test_save(self):
...
         assert team1 in Teams.objects.all()
+
+    def test_generate_teams(self, user_event_list, create_event):
+        for user_event in user_event_list:
+            assert user_event.teamID is None
+
+        Teams.generate_teams(create_event)
+        generated_user_event_list = UserEvent.objects.filter(eventID=create_event)
+        assert len(generated_user_event_list) == len(user_event_list)
+        team1 = []
```
Comment (ygalblum) [inline teams/tests.py:149]: Since you're just using the length of the arrays you can use counters instead of arrays.

File: teams/tests.py [hunk 5]
```diff
@@ -27,3 +138,21 @@ def test_save(self):
...
+        team2 = []
+        for user_event in generated_user_event_list:
+            if user_event.teamID.name == "Team1":
+                team1.append(user_event)
+            else:
+                team2.append(user_event)
+        team_size = len(generated_user_event_list) // 2
+        assert team_size == len(team1)
+        assert len(generated_user_event_list) - team_size == len(team2)
```
Comment (ygalblum) [inline teams/tests.py:158]: AFAICS, this `assert` is redundant.
You append to `team2` everyone that is not on `team1`. As a result, `len(team1) + len(team2) == generated_user_event_list`. Now, since you've already asseted that `team_size == len(team1)`, the second assert basically checks that `len(generated_user_event_list) - len(team1) == len(team2)` which is true by definition.
Right?

Comment (rotem2022) [reply teams/tests.py:158]: I believe both checks are necessary because of edge cases. For instance, let's consider the example where `len(generated_user_event_list)` = 5. So, 5//2 = 2. Therefore, `len(team1)` = 2 is correct, but it's possible that `len(team2) `could also be 2 instead of 3.

File: teams/tests.py (hunk 4 above)
Comment (rotem2022) [reply teams/tests.py:149]: yes I think you are right. thanks for the note.

File: teams/tests.py (hunk 3 above)
Comment (ygalblum) [reply teams/tests.py:7]: There's one django leftover

Comment (rotem2022) [reply teams/tests.py:7]: where is the leftover?

File: teams/tests.py [hunk 6]
```diff
@@ -1,13 +1,126 @@
...
+
 from django.core.exceptions import ValidationError
 from django.db.utils import IntegrityError
+from django.utils import timezone
+
+from event.models import UserEvent, Event
+from location.models import Location
+from category.models import Category
+from users.models import Profile
+from django.contrib.auth import get_user_model
```
Comment (ygalblum) [inline teams/tests.py:11]: Here it is :)

Comment (rotem2022) [reply teams/tests.py]: Thanks! changed (:

**Review by SaarOrus — APPROVED** [review]

**Review by talko611 — APPROVED** [review]

**Review by ygalblum — APPROVED** [review]
